parsl/executors/globus_compute.py

```diff
@@ -0,0 +1,139 @@
+    ):
+        """
+        Parameters
+        ----------
+
+        endpoint_id:
+            id of the endpoint to which to submit tasks
+
+        task_group_id:
+            The Task Group to which to associate tasks.  If not set,
+            one will be instantiated.
+
+        resource_specification:
+            Specify resource requirements for individual task execution.
+
+        user_endpoint_config:
+            User endpoint configuration values as described
+            and allowed by endpoint administrators. Must be a JSON-serializable dict
+            or None.
+
+        label:
+            a label to name the executor; mainly utilized for
+            logging and advanced needs with multiple executors.
+
+        batch_size:
+            the maximum number of tasks to coalesce before
+            sending upstream [min: 1, default: 128]
+
+        amqp_port:
+            Port to use when connecting to results queue. Note that the
+            Compute web services only support 5671, 5672, and 443.
+
+        kwargs:
+            Other kwargs listed will be passed through to globus_compute_sdk.Executor
+            as is
+        """
+        super().__init__()
+        self.endpoint_id = endpoint_id
+        self.task_group_id = task_group_id
+        self.resource_specification = resource_specification
+        self.user_endpoint_config = user_endpoint_config
+        self.label = label
+        self.batch_size = batch_size
+        self.amqp_port = amqp_port
+
+        try:
+            from globus_compute_sdk import Executor
+        except ImportError:
+            raise OptionalModuleMissing(
+                ['globus-compute-sdk'],
+                "GlobusComputeExecutor requires globus-compute-sdk installed"
+            )
+        self._executor: Executor = Executor(
+            endpoint_id=endpoint_id,
+            task_group_id=task_group_id,
+            resource_specification=resource_specification,
+            user_endpoint_config=user_endpoint_config,
+            label=label,
+            batch_size=batch_size,
+            amqp_port=amqp_port,
+            **kwargs
+        )
+
+    def start(self) -> None:
+        """Empty function
+        """
+        pass
+
+    def submit(self, func: Callable, resource_specification: Dict[str, Any], *args: Any, **kwargs: Any) -> Future:
+        """ Submit fn to globus-compute
+
+
+        Parameters
+        ----------
+
+        func: Callable
+            Python function to execute remotely
+        resource_specification: Dict[str, Any]
+            Resource specification used to run MPI applications on Endpoints configured
+            to use globus compute's MPIEngine
+        args:
+            Args to pass to the function
+        kwargs:
+            kwargs to pass to the function
+
+        Returns
+        -------
+
+        Future
+        """
+        self._executor.resource_specification = resource_specification or self.resource_specification
```

Comment:
if you want to specify that per task, it maybe makes sense to make it a new entry in the task submission resource_specification - that's the mechanism for specify everything about "how" a task is to be executed in Parsl

Reply:
Unless I'm misunderstanding something, that is how the multi-user endpoints are meant to work.  If you can't specify it at submit time, then the alternative will be to have multiple instantiations of `GlobusComputeExecutor` for the same endpoint id, but with different `user_endpoint_config`.  Maybe that's not a "bad thing", but while you would reduce number of endpoints to maintain, Parsl workflows would still have to have one `GlobusComputeExecutor` for each type of endpoint you want to interact with, even when on the same machine.  And maybe that's not so bad because even if you didn't have to do that, you'd still need to handle the multiple 'user_endpoint_config` dicts that are required for the various endpoint configs you need.  My thought was to match what is available on the Globus Compute side, which allows (I think) setting that at both Executor instantiation time and at submit time.

Reply:
i think per-task specification is a good option to make available. i was commenting on *where* that specification should go when submitting from Parsl.

Reply:
Would that look something like this?:
```
self._executor.user_endpoint_config = resource_specification["user_endpoint_config"] or self.user_endpoint_config
```

Reply:
yes

Reply:
I've made some changes to allow `resource_specification` and `user_endpoint_config` to both work.

```
# Specifying MPI apps
py_app(resource_specification={'num_nodes': 2, 'ranks_per_node':4})

# Specifying user_endpoint_config
py_app(resource_specification={'user_endpoint_config': {"account_id": "user_allocation_account_id"}}

# Specifying both:
py_app(resource_specification={'num_nodes': 2, 'ranks_per_node':4,
'user_endpoint_config': {"account_id": "user_allocation_account_id"}
})

```